ffmpeg_tools/validation.py

```diff
@@ -347,6 +359,47 @@ def validate_frame_rate(
     return True
 
 
+def validate_audio_sample_rates(
+        src_metadata: Dict[str, Any],
+        dest_audio_codec: str,
+        dst_audio_encoder_info: Dict[str, Any]) -> bool:
+
+    assert dest_audio_codec in codecs.AudioCodec._value2member_map_
+    assert codecs.AudioCodec(dest_audio_codec).get_encoder() is not None
+
+    if 'sample_rates' not in dst_audio_encoder_info:
+        # NOTE: `ffmpeg -h` unfortunately does not provide sample rates in all
+        # cases. The following encoders we support do not have this info:
+        # libopencore_amrnb, ac3, libvorbis, wmav2, pcm_u8.
+        #
+        # From emprical testing with ffmpeg I discovered the following:
+        # - pcm_u8 supports any value (tried up to 100 000 kHz).
+        # - libvorbis supports any value up to 200 kHz but crashes
+        #    on some unusual values (especially between 70 and 520 Hz).
+        # - wmav2 supports everything between 2 and 48 kHz.
+        # - ac3 supports only a few specific values, same as mp3.
+        # - libopencore_amrnb supports only 8 kHz.
```

Comment:
Sure. It's not a big change. I considered including it already but you originally rejected the idea of hard-coding sample rates if favor of `fffmpeg -h` and when that did not provide a complete solution you wanted to try with PyAV so I wasn't really sure if it would be acceptable to hard-code them now or if you would still prefer to seek some other ideas to automate it.

Reply:
Ah, so sorry for misunderstanding. I just proposed some solution that might work, but the goal was to block potential failures even if it means hardcoding.

Reply:
Done. Please see the top 4 commits:
1) I defined `codecs._SUPPORTED_SAMPLE_RATES` with the full set of known sample rates.
    - This includes those that `ffmpeg -h` does return. This is especially important because you said you won't be passing `encoder_info` to the validations and therefore you won't really get the rates from `ffmpeg -h` checked without this.
        - There's a small risk that it will go out of sync with what ffmpeg returns but I don't think this info really changes that often (if ever). And if you were worried about that you could provide `encoder_info`.
    - I decided to skip everything below 1000 Hz for `libvorbis`. I don't think anyone uses a rate that low in practice and this lets us conveniently step around the values that crashed ffmpeg in my experiments.
2) Validation logic has been updated:
    - If you do provide `encoder_info`, validation will try to use that.
    - If you don't provide it or if it does not contain the sample rates (i.e. `ffmpeg -h` did not list them), validation will fall back to using `_SUPPORTED_SAMPLE_RATES`.
    - If `_SUPPORTED_SAMPLE_RATES` does not list supported sample rates for your encoder, validation fails.
        - This is different than before. It used to pass when supported sample rates could not be determined simply because this was a common case. Now that `_SUPPORTED_SAMPLE_RATES` covers all supported audio encoders, we can safely fail in case someone manages to use an unsupported one (which is only possible if he skips target audio codec validations).
    - Note that sample rate validation is not (and cannot be) performed if you don't provide `muxer_info` and the target audio codec is not specified explicitly. This case is now possible after review changes from #23.
3) Sorry that it wasn't ready earlier but I wanted to do it properly rather than just hack it in:
    - Existing tests were updated and new ones added where necessary.
    - I implemented `is_supported_sample_rate()` in `codecs` + tests.
    - I needed to support both individual sample rates and ranges in `_SUPPORTED_SAMPLE_RATES`. I could just pretend everything is a range and do stuff like `[(48000, 48000), (44100, 44100), ...]` but that's not very DRY. I implemented a simple helper (`SparseRange`) that lets us define it in a more concise way and also provides a method for checking whether a specific sample rate falls into any of its sub-ranges.